lib/jsdom/living/helpers/shadow-dom.js

```diff
@@ -0,0 +1,252 @@
+"use strict";
+
+const NODE_TYPE = require("../node-type");
+const { domSymbolTree } = require("./internal-constants");
+
+const idlUtils = require("../generated/utils");
+
+// Valid host element for ShadowRoot.
+// Defined in: https://dom.spec.whatwg.org/#dom-element-attachshadow
+const VALID_HOST_ELEMENT_NAME = new Set([
+  "article",
+  "aside",
+  "blockquote",
+  "body",
+  "div",
+  "footer",
+  "h1",
+  "h2",
+  "h3",
+  "h4",
+  "h5",
+  "h6",
+  "header",
+  "main",
+  "nav",
+  "p",
+  "section",
+  "span"
+]);
+
+// TODO: Evaluate if it should be somewhere else.
+function isNode(nodeImpl) {
+  // Use an approximation by checking the presence of nodeType instead of instead of using the isImpl from
+  // "../generated/Node" to avoid introduction of circular dependencies.
+  return Boolean(nodeImpl && nodeImpl.nodeType);
```

Comment:
Maybe check `"nodeType" in nodeImpl` instead. This only works because nodeType is never 0, but I didn't know that until I checked the spec.